### ISSUE TITLE
Editor: Add "Upload code to Gist/Pastebin/hastebin..." feature

### ISSUE THREAD

**Issue by CaptainPRICE** [wiremod/wire#1239]
Sharing code is a good thing...
This feature would allow users to upload the code online, and share it faster.¹
- [GitHub Gist API](https://developer.github.com/v3/gists/)
  [About](https://en.wikipedia.org/wiki/GitHub#Gist). Permanent raw link: `https://gist.github.com/{gist_user}/{gist_id}/raw/{file_name}`
- [Pastebin API](http://pastebin.com/api)
  [About](http://pastebin.com/faq#1). See also [Scraping API](http://pastebin.com/api_scraping_faq).
- [Hastebin API](https://github.com/seejohnrun/haste-server/wiki/POST-api)
  [About](http://hastebin.com/about.md). Raw link: `http://hastebin.com/raw/{key}`

¹ _I would recommend to prompt the user to confirm their action, just in case._

**Comment by Divran** [comment]
We can't implement any such feature that requires the user to log in, because there is no way to hide the password from any server admin running malicious lua code.

**Comment by CaptainPRICE** [comment]
> can't implement any such feature that requires the user to log in

What the **** are you talking about?
GitHub Gist, Pastebin and hastebin... all allow public/anonymous uploads - no login/account is required.
What password?! Well there is API key.

**Comment by Divran** [comment]
I only meant that I am pretty sure that hastebin allows anonymous pastes, and that I wasn't sure about the others.

**Comment by bigdogmat** [comment]
I'm pretty sure both pastebin or github just require api keys to which on github I know you can restrict, not sure about pastebin.

**Comment by oldmud0** [comment]
How about OAuth2?

**Comment by Divran** [comment]
Ok and then how do you get the oauth key out of the browser and into the lua code so that we can upload the code using it?

**Comment by CaptainPRICE** [comment]
Moreover, there are "private" options for Gist and Pastebin.
GitHub calls it secret Gist. Pastebin calls it unlisted paste.
I am not sure about hastebin.
You simply never share the link if you want it "private". See, no user accounts required, etc.

**Comment by oldmud0** [comment]
The manual parts are required to keep a level of discretion/control by the end user. ShareX does something similar, where you have to copy-paste the OAuth key to the appropriate textbox for whatever integration.

**Comment by CaptainPRICE** [comment]
> A simple anonymous "Upload to Pastebin" plus a couple of other well-established services is fine

Yes. This. No need for user accounts/logins, passwords, etc... I have never mentioned to do the whole thing with user accounts, and so on.
Ok, let's just do public and private/anonymous (secret/unlisted) sharing, that does not require user accounts/passwords.. It is a simple HTTP (POST) request.
So... Focus on it. **API key.**
If wire comes with builtin API key, chances are, Pastebin will block it due too many requests or something... Well, I think it takes just five players to keep spamming (POSTing) HTTP requests repeatedly (with same API key) for it to be blocked by Pastebin after a couple of minutes. I am not sure about Gist; never tried it..
There could also be a daily/weekly/monthly limit or something.
It is all on the client-side, nothing is networked to the server-side. So, the question is, how do you save API key on the client, without server being able to see it? Because, it would be very annoying to have to enter API key every single time. I have an idea, here it is:
When the user clicks `Upload to Gist/Pastebin/hastebin` button, it would check if there is a file `key.txt`, if this is the first time the user clicked it - it will return false. So, the user is prompted to enter API key, and their secret phrase (API key would be encrypted with user's secret phrase). Next time the user clicks `Upload to Gist/Pastebin/hastebin` button, `key.txt` file will exist, so it would return true, and this time it would only ask a user to enter their secret phrase.
This way, only "infected"/bad servers could reach to client's file system, and even if they do, it will be encrypted, so if they doesn't know the secret phrase, they could try brute-force technique, but nothing can stop "bad" hackers....
So, what do you think of that approach?

**Comment by oldmud0** [comment]
If it requires an API key inputted manually then forget it.

**Comment by CaptainPRICE** [comment]
> Gmod does have javascript integration, so I guess that could work, but that can be intercepted by admin abusers.

Correction: Anything can be intercepted _only_ if wire has been infected, such as a client sending such information to the server and then to admin abuser...
Let's not worry about that, because that's again the story about players' trust the server they are playing on...

So, **without API key**. Okay. @oldmud0 has already explained it a bit. Now I will try to round it.
When the user clicks `Upload to Gist/Pastebin/hastebin` button, it would copy the code to clipboard, open Steam web browser in overlay, and automatically navigate to the wanted service (see `gui.OpenURL` function). But, there is a problem with that, Steam overlay can be enabled/disabled by the user via Steam settings. So, instead of using Steam overlay, create a new window/frame with [DHTML](https://wiki.garrysmod.com/page/Category:DHTML) on it, and use JavaScript integration to automate it as much as possible. This way, a user can be presented with website interface, allowing them to set any options offered by services such as Gist/Pastebin/hastebin. Minimizing the number of clicks the user has to do, right now I think it can be minimized down to only 2 clicks.
What are your thoughts about this approach?

**Comment by AbigailBuccaneer** [comment]
If the text editor had a useful integration point for third-party addons (eg. calling a hook and passing in the menu bar as an action), this could easily be achieved by a third-party addon. It would feel like serious feature creep for Wiremod to be responsible for dealing with API keys or manipulating third-party websites via Javascript injection.

**Comment by CaptainPRICE** [comment]
> If the text editor had a useful integration point for third-party addons

It sounds like Editor was not designed and constructed with extensibility in mind.
So, new issue?

> It would feel like serious feature creep for Wiremod to be responsible...

Yeah, I agree..
